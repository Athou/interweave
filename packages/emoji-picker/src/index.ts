/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { EmojiPicker } from './Picker';

export * from './constants';
export * from './types';

// Avoid exporting internal picker
export { EmojiPicker };
