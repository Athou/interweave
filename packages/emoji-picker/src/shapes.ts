/**
 * @copyright   2016-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import React from 'react';
import PropTypes from 'prop-types';

// eslint-disable-next-line import/prefer-default-export
export const ContextShape = PropTypes.shape({
  classNames: PropTypes.objectOf(PropTypes.string).isRequired,
  messages: PropTypes.objectOf(PropTypes.node).isRequired,
});
