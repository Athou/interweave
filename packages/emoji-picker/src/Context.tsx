/**
 * @copyright   2016-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import React from 'react';
import { ContextShape } from './shapes';
import { Context as EmojiContext } from './types';

export const Context = React.createContext({
  classNames: {},
  messages: {},
});

Context.Provider.propTypes = {
  value: ContextShape.isRequired,
};

export interface WithContextProps {
  context: EmojiContext;
}

export default function withContext<Props extends {}>(
  Component: React.ComponentType<Props & WithContextProps>,
) {
  return function WithContextWrapper(props: Props) {
    return (
      <Context.Consumer>{context => <Component {...props} context={context} />}</Context.Consumer>
    );
  };
}
