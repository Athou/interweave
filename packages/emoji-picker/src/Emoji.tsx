/**
 * @copyright   2016-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import React from 'react';
import PropTypes from 'prop-types';
import EmojiCharacter, {
  CanonicalEmoji,
  EmojiShape,
  Path,
  PathShape,
  Source,
  SourceShape,
} from 'interweave-emoji';
import withContext, { ContextProps } from './Context';
import { ContextShape } from './shapes';

export interface EmojiProps {
  active: boolean;
  emoji: CanonicalEmoji;
  emojiPadding: number;
  emojiPath: Path;
  emojiSize: number;
  emojiSource: Source;
  onEnter: (emoji: CanonicalEmoji, event: React.MouseEvent<HTMLButtonElement>) => void;
  onLeave: (emoji: CanonicalEmoji, event: React.MouseEvent<HTMLButtonElement>) => void;
  onSelect: (emoji: CanonicalEmoji, event: React.MouseEvent<HTMLButtonElement>) => void;
  showImage: boolean;
}

export interface EmojiState {
  active: boolean;
}

export class Emoji extends React.PureComponent<EmojiProps & ContextProps, EmojiState> {
  static propTypes = {
    active: PropTypes.bool.isRequired,
    context: ContextShape.isRequired,
    emoji: EmojiShape.isRequired,
    emojiPadding: PropTypes.number.isRequired,
    emojiPath: PathShape.isRequired,
    emojiSize: PropTypes.number.isRequired,
    emojiSource: SourceShape.isRequired,
    onEnter: PropTypes.func.isRequired,
    onLeave: PropTypes.func.isRequired,
    onSelect: PropTypes.func.isRequired,
    showImage: PropTypes.bool.isRequired,
  };

  state = {
    active: this.props.active,
  };

  componentDidUpdate(prevProps: EmojiProps) {
    if (this.props.active !== prevProps.active) {
      this.setState({
        active: this.props.active,
      });
    }
  }

  /**
   * Triggered when the emoji is clicked.
   */
  private handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();

    this.props.onSelect(this.props.emoji, event);
  };

  /**
   * Triggered when the emoji is hovered.
   */
  private handleEnter = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();

    this.setState({
      active: true,
    });

    this.props.onEnter(this.props.emoji, event);
  };

  /**
   * Triggered when the emoji is no longer hovered.
   */
  private handleLeave = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();

    this.setState({
      active: false,
    });

    this.props.onLeave(this.props.emoji, event);
  };

  render() {
    const {
      context: { classNames },
      emoji,
      emojiPadding,
      emojiPath,
      emojiSize,
      emojiSource,
      showImage,
    } = this.props;
    const { active } = this.state;
    const dimension = emojiPadding + emojiPadding + emojiSize;
    const className = [classNames.emoji];

    if (active) {
      className.push(classNames.emojiActive);
    }

    return (
      <button
        aria-label={emoji.annotation}
        key={emoji.hexcode}
        className={className.join(' ')}
        style={{
          height: dimension,
          padding: emojiPadding,
          width: dimension,
        }}
        type="button"
        onClick={this.handleClick}
        onMouseEnter={this.handleEnter}
        onMouseLeave={this.handleLeave}
      >
        {showImage ? (
          <EmojiCharacter
            emojiPath={emojiPath}
            emojiSize={emojiSize}
            emojiSource={emojiSource}
            hexcode={emoji.hexcode}
          />
        ) : (
          <div
            style={{
              height: emojiSize,
              overflow: 'hidden',
              visibility: 'hidden',
              width: emojiSize,
            }}
          >
            &nbsp;
          </div>
        )}
      </button>
    );
  }
}

export default withContext(Emoji);
