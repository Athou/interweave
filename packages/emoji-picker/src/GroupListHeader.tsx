/**
 * @copyright   2016-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import React from 'react';
import PropTypes from 'prop-types';
import camelCase from 'lodash/camelCase';
import withContext, { ContextProps } from './Context';
import {
  GROUP_KEY_COMMONLY_USED,
  GROUP_KEY_SEARCH_RESULTS,
  GROUP_KEY_SMILEYS_PEOPLE,
  GROUP_KEY_NONE,
} from './constants';
import { ContextShape } from './shapes';
import { CommonMode, GroupKey } from './types';

export interface GroupListHeaderProps {
  commonMode: CommonMode;
  group: GroupKey;
  skinTonePalette?: React.ReactNode;
}

export class GroupListHeader extends React.PureComponent<GroupListHeaderProps & ContextProps> {
  static propTypes = {
    commonMode: PropTypes.string.isRequired,
    context: ContextShape.isRequired,
    group: PropTypes.string.isRequired,
    skinTonePalette: PropTypes.node,
  };

  static defaultProps = {
    skinTonePalette: null,
  };

  render() {
    const {
      commonMode,
      context: { classNames, messages },
      group,
      skinTonePalette,
    } = this.props;
    const showPalette =
      skinTonePalette &&
      (group === GROUP_KEY_SMILEYS_PEOPLE ||
        group === GROUP_KEY_SEARCH_RESULTS ||
        group === GROUP_KEY_NONE);

    return (
      <header className={classNames.emojisHeader}>
        {showPalette && skinTonePalette}

        {group === GROUP_KEY_COMMONLY_USED
          ? messages[camelCase(commonMode)]
          : messages[camelCase(group)]}
      </header>
    );
  }
}

export default withContext(GroupListHeader);
