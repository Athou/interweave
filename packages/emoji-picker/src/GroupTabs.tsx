/**
 * @copyright   2016-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import React from 'react';
import PropTypes from 'prop-types';
import camelCase from 'lodash/camelCase';
import { EmojiShape, CanonicalEmoji } from 'interweave-emoji';
import withContext, { WithContextProps } from './Context';
import Group from './Group';
import { GROUPS, GROUP_KEY_COMMONLY_USED, GROUP_ICONS } from './constants';
import { ContextShape } from './shapes';
import { CommonMode, GroupKey } from './types';

export interface GroupTabsProps {
  activeGroup: GroupKey;
  commonEmojis: CanonicalEmoji[];
  commonMode: CommonMode;
  icons: { [key: string]: React.ReactNode };
  onSelect: (group: GroupKey, event: React.MouseEvent<HTMLButtonElement>) => void;
}

export class GroupTabs extends React.PureComponent<GroupTabsProps & WithContextProps> {
  static propTypes = {
    activeGroup: PropTypes.string.isRequired,
    commonEmojis: PropTypes.arrayOf(EmojiShape).isRequired,
    commonMode: PropTypes.string.isRequired,
    context: ContextShape.isRequired,
    icons: PropTypes.objectOf(PropTypes.node).isRequired,
    onSelect: PropTypes.func.isRequired,
  };

  render() {
    const {
      activeGroup,
      commonEmojis,
      commonMode,
      context: { classNames },
      icons,
      onSelect,
    } = this.props;
    const groups = [...GROUPS];

    if (commonEmojis.length > 0) {
      groups.unshift(GROUP_KEY_COMMONLY_USED);
    }

    return (
      <nav className={classNames.groups}>
        <ul className={classNames.groupsList}>
          {groups.map(group => (
            <li key={group}>
              <Group
                activeGroup={activeGroup}
                commonMode={commonMode}
                group={group}
                onSelect={onSelect}
              >
                {icons[group] || icons[camelCase(group)] || GROUP_ICONS[group]}
              </Group>
            </li>
          ))}
        </ul>
      </nav>
    );
  }
}

export default withContext(GroupTabs);
