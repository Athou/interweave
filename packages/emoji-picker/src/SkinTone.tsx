/**
 * @copyright   2016-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import React from 'react';
import PropTypes from 'prop-types';
import camelCase from 'lodash/camelCase';
import withContext, { WithContextProps } from './withContext';
import { SKIN_COLORS } from './constants';
import { ContextShape } from './shapes';
import { SkinToneKey } from './types';

export interface SkinToneProps {
  active: boolean;
  children?: React.ReactNode;
  onSelect: (skinTone: SkinToneKey, event: React.MouseEvent<HTMLButtonElement>) => void;
  skinTone: SkinToneKey;
}

export class SkinTone extends React.PureComponent<SkinToneProps & WithContextProps> {
  static propTypes = {
    active: PropTypes.bool.isRequired,
    children: PropTypes.node,
    context: ContextShape.isRequired,
    onSelect: PropTypes.func.isRequired,
    skinTone: PropTypes.string.isRequired,
  };

  static defaultProps = {
    children: null,
  };

  /**
   * Triggered when the button is clicked.
   */
  private handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();

    this.props.onSelect(this.props.skinTone, event);
  };

  render() {
    const {
      active,
      children,
      context: { classNames, messages },
      skinTone,
    } = this.props;
    const className = [classNames.skinTone];
    const color = SKIN_COLORS[skinTone];
    const key = camelCase(skinTone);

    if (active) {
      className.push(classNames.skinToneActive);
    }

    return (
      <button
        className={className.join(' ')}
        style={{ backgroundColor: color, borderColor: color, color }}
        title={messages[`skin${key.charAt(0).toUpperCase()}${key.slice(1)}`]}
        type="button"
        onClick={this.handleClick}
      >
        {children || ' '}
      </button>
    );
  }
}

export default withContext(SkinTone);
