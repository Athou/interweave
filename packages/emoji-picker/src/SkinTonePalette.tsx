/**
 * @copyright   2016-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import React from 'react';
import PropTypes from 'prop-types';
import SkinTone from './SkinTone';
import withContext, { ContextProps } from './Context';
import { SKIN_TONES } from './constants';
import { ContextShape } from './shapes';
import { Context, SkinToneKey } from './types';

export interface SkinTonePaletteProps {
  activeSkinTone: SkinToneKey;
  icons: { [key: string]: React.ReactNode };
  onSelect: (skinTone: SkinToneKey, event: React.MouseEvent<HTMLButtonElement>) => void;
}

export class SkinTonePalette extends React.PureComponent<SkinTonePaletteProps & ContextProps> {
  static propTypes = {
    activeSkinTone: PropTypes.string.isRequired,
    context: ContextShape.isRequired,
    icons: PropTypes.objectOf(PropTypes.node).isRequired,
    onSelect: PropTypes.func.isRequired,
  };

  render() {
    const {
      activeSkinTone,
      context: { classNames },
      icons,
      onSelect,
    } = this.props;

    return (
      <div className={classNames.skinTones}>
        {SKIN_TONES.map(skinTone => (
          <SkinTone
            key={skinTone}
            active={activeSkinTone === skinTone}
            skinTone={skinTone}
            onSelect={onSelect}
          >
            {icons[skinTone] || null}
          </SkinTone>
        ))}
      </div>
    );
  }
}

export default withContext(SkinTonePalette);
