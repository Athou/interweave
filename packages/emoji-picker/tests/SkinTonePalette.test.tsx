import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { SKIN_TONES, SKIN_KEY_NONE, SKIN_KEY_DARK, SKIN_KEY_MEDIUM_LIGHT } from '../src/constants';
import SkinTone from '../src/SkinTone';
import { SkinTonePalette, SkinTonePaletteProps } from '../src/SkinTonePalette';
import { WithContextProps } from '../src/withContext';
import { PICKER_CONTEXT } from './mocks';

function findSkinToneByKey(wrapper: ShallowWrapper, key: string) {
  return wrapper
    .find('li')
    .filterWhere(node => node.key() === key)
    .find(SkinTone)
    .at(0);
}

describe('<SkinTonePalette />', () => {
  const props: SkinTonePaletteProps & WithContextProps = {
    activeSkinTone: SKIN_KEY_NONE,
    context: PICKER_CONTEXT,
    icons: {},
    onSelect() {},
  };

  it('renders a palette', () => {
    const wrapper = shallow(<SkinTonePalette {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders a node for each skin tone', () => {
    const wrapper = shallow(<SkinTonePalette {...props} />);

    expect(wrapper.find(SkinTone)).toHaveLength(SKIN_TONES.length);
  });

  it('can customize class name', () => {
    const wrapper = shallow(
      <SkinTonePalette
        {...props}
        context={{
          ...props.context,
          classNames: {
            ...props.context.classNames,
            skinTones: 'test-skin-tones',
          },
        }}
      />,
    );

    expect(wrapper.prop('className')).toBe('test-skin-tones');
  });

  it('passes `onSelect` to skin tones', () => {
    const wrapper = shallow(<SkinTonePalette {...props} />);

    expect(
      wrapper
        .find(SkinTone)
        .at(0)
        .prop('onSelect'),
    ).toBe(props.onSelect);
  });

  it('sets active skin tone', () => {
    const wrapper = shallow(<SkinTonePalette {...props} activeSkinTone={SKIN_KEY_DARK} />);

    expect(findSkinToneByKey(wrapper, SKIN_KEY_NONE).prop('active')).toBe(false);
    expect(findSkinToneByKey(wrapper, SKIN_KEY_DARK).prop('active')).toBe(true);
  });

  it('renders default icon as null', () => {
    const wrapper = shallow(<SkinTonePalette {...props} />);

    expect(findSkinToneByKey(wrapper, SKIN_KEY_MEDIUM_LIGHT).prop('children')).toBeNull();
  });

  it('renders icon using kebab case', () => {
    const icon = <b>Animal</b>;
    const wrapper = shallow(<SkinTonePalette {...props} icons={{ 'medium-light': icon }} />);

    expect(findSkinToneByKey(wrapper, SKIN_KEY_MEDIUM_LIGHT).contains(icon)).toBe(true);
  });

  it('renders icon using camel case', () => {
    const icon = <b>Animal</b>;
    const wrapper = shallow(<SkinTonePalette {...props} icons={{ mediumLight: icon }} />);

    expect(findSkinToneByKey(wrapper, SKIN_KEY_MEDIUM_LIGHT).contains(icon)).toBe(true);
  });
});
