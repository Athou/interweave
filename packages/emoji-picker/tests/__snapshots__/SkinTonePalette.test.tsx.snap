// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkinTonePalette renders a palette 1`] = `
<nav className="interweave-picker__skin-tones">
  <ul className="interweave-picker__skin-tones-list">
    <li>
      <SkinTone active skinTone="none" onSelect={onSelect()}>
        <button
          className="interweave-picker__skin-tone interweave-picker__skin-tone--active"
          data-skin-color="#FFCC22"
          data-skin-tone="none"
          style={{ backgroundColor: "#FFCC22", borderColor: "#FFCC22", color: "#FFCC22" }}
          title="No skin tone"
          type="button"
          onClick={handleClick()}
        >
           
        </button>
      </SkinTone>
    </li>
    <li>
      <SkinTone active={false} skinTone="light" onSelect={onSelect()}>
        <button
          className="interweave-picker__skin-tone"
          data-skin-color="#FADCBC"
          data-skin-tone="light"
          style={{ backgroundColor: "#FADCBC", borderColor: "#FADCBC", color: "#FADCBC" }}
          title="Light skin tone"
          type="button"
          onClick={handleClick()}
        >
           
        </button>
      </SkinTone>
    </li>
    <li>
      <SkinTone active={false} skinTone="medium-light" onSelect={onSelect()}>
        <button
          className="interweave-picker__skin-tone"
          data-skin-color="#E0BB95"
          data-skin-tone="medium-light"
          style={{ backgroundColor: "#E0BB95", borderColor: "#E0BB95", color: "#E0BB95" }}
          title="Medium-light skin tone"
          type="button"
          onClick={handleClick()}
        >
           
        </button>
      </SkinTone>
    </li>
    <li>
      <SkinTone active={false} skinTone="medium" onSelect={onSelect()}>
        <button
          className="interweave-picker__skin-tone"
          data-skin-color="#BF8F68"
          data-skin-tone="medium"
          style={{ backgroundColor: "#BF8F68", borderColor: "#BF8F68", color: "#BF8F68" }}
          title="Medium skin tone"
          type="button"
          onClick={handleClick()}
        >
           
        </button>
      </SkinTone>
    </li>
    <li>
      <SkinTone active={false} skinTone="medium-dark" onSelect={onSelect()}>
        <button
          className="interweave-picker__skin-tone"
          data-skin-color="#9B643D"
          data-skin-tone="medium-dark"
          style={{ backgroundColor: "#9B643D", borderColor: "#9B643D", color: "#9B643D" }}
          title="Medium-dark skin tone"
          type="button"
          onClick={handleClick()}
        >
           
        </button>
      </SkinTone>
    </li>
    <li>
      <SkinTone active={false} skinTone="dark" onSelect={onSelect()}>
        <button
          className="interweave-picker__skin-tone"
          data-skin-color="#5A463A"
          data-skin-tone="dark"
          style={{ backgroundColor: "#5A463A", borderColor: "#5A463A", color: "#5A463A" }}
          title="Dark skin tone"
          type="button"
          onClick={handleClick()}
        >
           
        </button>
      </SkinTone>
    </li>
  </ul>
</nav>
`;
